### PR TITLE
Add upsert profile and track event actions to klaviyo

### PR DESCRIPTION
**Pull request by harsh-joshi99**
This PR adds the Upsert profile, Track Event and Order completed actions to klaviyo destination.
JIRA -> 
          [STRATCONN-3165](https://segment.atlassian.net/browse/STRATCONN-3165),
          [STRATCONN-3164](https://segment.atlassian.net/browse/STRATCONN-3164),
          [STRATCONN-3163](https://segment.atlassian.net/browse/STRATCONN-3163)

## Testing
Tested in local and stage.

Stage testing results -> https://docs.google.com/document/d/1k9sFVNKiyrR5dhzpcZ3ChJizUXSVMkntgfIlCsVKg9g/edit?usp=sharing
_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [x] [Segmenters] Tested in the staging environment

[STRATCONN-3165]: https://segment.atlassian.net/browse/STRATCONN-3165?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[STRATCONN-3164]: https://segment.atlassian.net/browse/STRATCONN-3164?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[STRATCONN-3163]: https://segment.atlassian.net/browse/STRATCONN-3163?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/destination-actions/src/destinations/klaviyo/index.ts [hunk 1]
```diff
@@ -51,13 +54,15 @@ const destination: DestinationDefinition<Settings> = {
       headers: {
         Authorization: `Klaviyo-API-Key ${settings.api_key}`,
         Accept: 'application/json',
-        revision: new Date().toISOString().slice(0, 10)
+        revision: '2023-09-15'
       }
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/klaviyo/index.ts:58]: Could you extract this to ./config as well?

File: packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts [hunk 2]
```diff
@@ -2,18 +2,175 @@ import type { ActionDefinition } from '@segment/actions-core'
...
+      }
     }
   },
-  perform: () => {
-    return undefined
+  perform: async (request, { payload }) => {
+    const { email, external_id, phone_number, ...otherAttributes } = payload
+
+    if (!email && !phone_number && !external_id) {
+      throw new IntegrationError(
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts:129]: Could you change this to  PayloadValidationError?

File: packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts [hunk 3]
```diff
@@ -2,18 +2,175 @@ import type { ActionDefinition } from '@segment/actions-core'
...
+    },
+    phone_number: {
+      label: 'Phone Number',
+      description: `Individual's phone number in E.164 format. If SMS is not enabled and if you use Phone Number as identifier, then you have to provide one of Email or External ID.`,
+      type: 'string',
+      default: { '@path': '$.context.traits.phone' }
+    },
+    external_id: {
+      label: 'External ID',
+      description: `A unique identifier used by customers to associate Klaviyo profiles with profiles in an external system.`,
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts:29]: Could you update the description here to indicate one of Email, Phone Number or external ID?

File: packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts [hunk 4]
```diff
@@ -2,18 +2,175 @@ import type { ActionDefinition } from '@segment/actions-core'
...
+
+          const profile = await request(`${API_URL}/profiles/${id}`, {
+            method: 'PATCH',
+            json: profileData
+          })
+          return profile
+        }
+      }
+
+      throw new Error('An error occurred while processing the request')
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts:172]: Could you use of the predefined error classes?

File: packages/destination-actions/src/destinations/klaviyo/trackEvent/index.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    unique_id: {
+      label: 'Unique ID',
+      description: `A unique identifier for an event. If the unique_id is repeated for the same
+      profile and metric, only the first processed event will be recorded. If this is not
+      present, this will use the time to the second. Using the default, this limits only one
+      event per profile per second.
+      `,
+      type: 'string',
+      default: {
+        '@path': '$.event'
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/klaviyo/trackEvent/index.ts:79]: Should this be mapped to $.messageId given `metric_name` is $.event?

File: packages/destination-actions/src/destinations/klaviyo/trackEvent/index.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+      }
+    }
+  },
+  perform: (request, { payload }) => {
+    const { email, phone_number } = payload.profile
+
+    if (!email && !phone_number) {
+      throw new IntegrationError('One of Phone Number or Email is required.', 'Missing required fields', 400)
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/klaviyo/trackEvent/index.ts:87]: Pls change to PayloadValidationError

**Review by varadarajan-tw — COMMENTED** [review]
Looks Great! ✨ Left few minor comments.

File: packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts [hunk 7]
```diff
@@ -2,18 +2,171 @@ import type { ActionDefinition } from '@segment/actions-core'
...
+
+          const profile = await request(`${API_URL}/profiles/${id}`, {
+            method: 'PATCH',
+            json: profileData
+          })
+          return profile
+        }
+      }
+
+      throw new APIError('An error occurred while processing the request', 400)
```
Comment (varadarajan-tw) [inline packages/destination-actions/src/destinations/klaviyo/upsertProfile/index.ts:168]: Do you want to indicate in the error message that Update Failed?

**Review by varadarajan-tw — APPROVED** [review]

**Comment by joe-ayoub-segment** [comment]
hi @harsh-joshi99 this PR has been deployed

**Pull request merged**